File: src/labels.py
"""Portable workspace label → colour rules.

Chromatic itself stored identities by workspace id. Personal setups also keep a
version-1 JSON map of labels to colours (chromatic-spaces-identities.json). This
module reads that map and applies it to matching live labels without writing the
source file.
"""

import json
import os

import ctx
import identity as ident
import state as state_mod


class LabelRulesError(ValueError):
    pass


def _legacy_identities_path():
    return os.path.join(
        os.path.dirname(ctx.herdr_config_path()),
        "chromatic-spaces-identities.json",
    )


def rules_path():
    env = os.environ.get("HERDR_LABEL_IDENTITIES_FILE")
    if env:
        return env
    settings = ctx.settings()
    configured = settings.get("label_identities_file")
    if configured:
        return configured
    plugin_path = os.path.join(ctx.config_dir(), "identities.json")
    if os.path.exists(plugin_path):
        return plugin_path
    legacy = _legacy_identities_path()
    if os.path.exists(legacy):
        return legacy
    return plugin_path


def _load_file(path):
    if not path or not os.path.exists(path):
        return {}
    try:
        with open(path, "r", encoding="utf-8") as fh:
            data = json.load(fh)
    except (ValueError, OSError) as exc:
        raise LabelRulesError("unreadable label identities file %s: %s" % (path, exc))
    if not isinstance(data, dict) or data.get("version") != 1:
        raise LabelRulesError(
            "label identities %s must be version 1 with an identities object" % path
        )
    raw = data.get("identities")
    if not isinstance(raw, dict):
        raise LabelRulesError("label identities %s missing identities object" % path)
    out = {}
    for label, colour in raw.items():
        if not isinstance(label, str) or not label or not isinstance(colour, str) or not colour:
            raise LabelRulesError("invalid label identity entry in %s" % path)
        resolved = ident.resolve_colour(colour)
        if not resolved:
            raise LabelRulesError(
                "label %r colour %r is not a palette name or #rrggbb" % (label, colour)
            )
        out[label] = resolved
    return out


def load_rules():
    """File rules first, then optional settings overlay."""
    rules = _load_file(rules_path())
    inline = ctx.settings().get("label_identities") or {}
    if not isinstance(inline, dict):
        raise LabelRulesError("settings.label_identities must be an object")
    for label, colour in inline.items():
        if not isinstance(label, str) or not label:
            continue
        resolved = ident.resolve_colour(colour) if isinstance(colour, str) else None
        if not resolved:
            ctx.warn("ignoring invalid settings label identity %r -> %r" % (label, colour))
            continue
        rules[label] = resolved
    return rules


def apply(st, workspaces):
    """Assign colours for matching labels unless the user set them manually.

    Returns workspace ids that changed.
    """
    try:
        rules = load_rules()
    except LabelRulesError as exc:
        ctx.warn(str(exc))
        return []
    if not rules:
        return []
    changed = []
    for w in workspaces or []:
        wid = w.get("workspace_id")
        label = w.get("label")
        if not wid or not label or label not in rules:
            continue
        colour = rules[label]
        current = state_mod.identity_of(st, wid) or {}
        if current.get("origin") == "manual":
            continue
        if current.get("colour") == colour and current.get("origin") == "label":
            continue
        state_mod.set_identity(st, wid, colour, origin="label")
        changed.append(wid)
    return changed

File: src/spaces/labels.ts
import { existsSync, readFileSync } from "node:fs"
import { dirname, join } from "node:path"

import { Predicate, Result, Schema } from "effect"

import { LabelRulesError } from "../runtime/errors.ts"
import type { PluginPathValues } from "../runtime/paths.ts"
import { loadSettings, settingsString, type PluginSettings } from "../runtime/settings.ts"
import { setIdentity, type PluginState, identityOf } from "../state/store.ts"
import { resolveColour, type WorkspaceSnapshot } from "./identity.ts"

type Json = typeof Schema.Json.Type

export type LabelRuleSet = ReadonlyArray<readonly [string, string]>

export type InvalidInlineRule = {
  readonly label: string
  readonly colour: Json
}

export type LabelApplyOutcome = {
  readonly changed: string[]
  readonly error?: LabelRulesError
}

export function rulesPath(
  paths: PluginPathValues,
  settings: PluginSettings,
  envFile?: string,
): string {
  if (envFile !== undefined && envFile !== "") return envFile

  const configured = settingsString(settings, "label_identities_file")

  if (configured !== undefined) return configured

  const pluginPath = join(paths.configDir, "identities.json")

  if (existsSync(pluginPath)) return pluginPath

  const legacy = join(dirname(paths.herdrConfigPath), "chromatic-spaces-identities.json")

  if (existsSync(legacy)) return legacy

  return pluginPath
}

export function parseLabelFile(data: Json, path: string): LabelRuleSet {
  const decoded = Schema.decodeUnknownResult(Schema.JsonObject)(data)

  if (Result.isFailure(decoded) || decoded.success.version !== 1) {
    throw new LabelRulesError({
      message: `label identities ${path} must be version 1 with an identities object`,
    })
  }

  const raw = Schema.decodeUnknownResult(Schema.JsonObject)(decoded.success.identities ?? null)

  if (Result.isFailure(raw)) {
    throw new LabelRulesError({
      message: `label identities ${path} missing identities object`,
    })
  }

  const out: Array<readonly [string, string]> = []

  for (const label of Object.keys(raw.success)) {
    const colour = raw.success[label]

    if (label === "" || !Predicate.isString(colour) || colour === "") {
      throw new LabelRulesError({
        message: `invalid label identity entry in ${path}`,
      })
    }

    const resolved = resolveColour(colour)

    if (resolved === undefined) {
      throw new LabelRulesError({
        message: `label ${JSON.stringify(label)} colour ${JSON.stringify(colour)} is not a palette name or #rrggbb`,
      })
    }

    out.push([label, resolved])
  }

  return out
}

function loadFile(path: string): LabelRuleSet {
  if (path === "" || !existsSync(path)) return []

  let parsed: unknown

  try {
    parsed = JSON.parse(readFileSync(path, "utf8"))
  } catch (cause) {
    const detail = cause instanceof Error ? cause.message : String(cause)

    throw new LabelRulesError({
      message: `unreadable label identities file ${path}: ${detail}`,
    })
  }

  const json = Schema.decodeUnknownResult(Schema.Json)(parsed)

  if (Result.isFailure(json)) {
    throw new LabelRulesError({
      message: `unreadable label identities file ${path}: not JSON`,
    })
  }

  return parseLabelFile(json.success, path)
}

function overlayInline(
  rules: Map<string, string>,
  inline: Json,
  onInvalid: (entry: InvalidInlineRule) => void,
): void {
  const decoded = Schema.decodeUnknownResult(Schema.JsonObject)(inline)

  if (Result.isFailure(decoded)) {
    throw new LabelRulesError({
      message: "settings.label_identities must be an object",
    })
  }

  for (const label of Object.keys(decoded.success)) {
    if (label === "") continue

    const colour = decoded.success[label]
    const resolved = Predicate.isString(colour) ? resolveColour(colour) : undefined

    if (resolved === undefined) {
      onInvalid({ label, colour: colour ?? null })
      continue
    }

    rules.set(label, resolved)
  }
}

export function loadRules(
  paths: PluginPathValues,
  settings: PluginSettings = loadSettings(paths),
  envFile = process.env.HERDR_LABEL_IDENTITIES_FILE,
): LabelRuleSet {
  const rules = new Map(loadFile(rulesPath(paths, settings, envFile)))
  const inline = settings.label_identities ?? {}

  overlayInline(rules, inline, () => {
    // caller logs invalid overlay entries
  })

  return [...rules.entries()]
}

export function loadRulesWithWarn(
  paths: PluginPathValues,
  onInvalidInline: (entry: InvalidInlineRule) => void,
  settings: PluginSettings = loadSettings(paths),
  envFile = process.env.HERDR_LABEL_IDENTITIES_FILE,
): LabelRuleSet {
  const rules = new Map(loadFile(rulesPath(paths, settings, envFile)))

  overlayInline(rules, settings.label_identities ?? {}, onInvalidInline)

  return [...rules.entries()]
}

export function applyLabelRules(
  state: PluginState,
  workspaces: readonly WorkspaceSnapshot[],
  rules: LabelRuleSet,
): string[] {
  if (rules.length === 0) return []

  const byLabel = new Map(rules)
  const changed: string[] = []

  for (const workspace of workspaces) {
    const wid = workspace.workspace_id
    const label = workspace.label

    if (wid === undefined || wid === "" || label === undefined || label === "") continue

    const colour = byLabel.get(label)

    if (colour === undefined) continue

    const current = identityOf(state, wid) ?? {}

    if (current.origin === "manual") continue

    if (current.colour === colour && current.origin === "label") continue

    setIdentity(state, wid, colour, "label")
    changed.push(wid)
  }

  return changed
}

export function tryApplyLabelRules(
  state: PluginState,
  workspaces: readonly WorkspaceSnapshot[],
  paths: PluginPathValues,
): LabelApplyOutcome {
  try {
    const changed = applyLabelRules(state, workspaces, loadRules(paths))

    return { changed } satisfies LabelApplyOutcome
  } catch (cause) {
    if (cause instanceof LabelRulesError) {
      return { changed: [], error: cause } satisfies LabelApplyOutcome
    }

    throw cause
  }
}
